pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go

```diff
@@ -73,6 +77,12 @@ type ResolvedPipelineRunTask struct {
 	ResolvedTaskResources *resources.ResolvedTaskResources
 	// ConditionChecks ~~TaskRuns but for evaling conditions
 	ResolvedConditionChecks TaskConditionCheckState // Could also be a TaskRun or maybe just a Pod?
+	// IsFailurePermitted is driven by the runOn section of the dependent tasks, runOn:  state: ["failure"]
+	// it is set to true if any of the tasks within the pipeline has dependency on this task's failure state
+	// by default, IsFailurePermitted is set to false
+	IsFailurePermitted bool
```

Comment:
I would like to propose a change in how we handle this logic here.
I'm not sure I've thought enough about this, but these are the basic ideas:
- any taskrun is "permitted" to fail, it simply fails
- we include failed taskruns in the list of executed tasks
- we include the filters defined by the `runOn` in the logic that calculates which task should be included in the list of schedulable ones at a certain point in time
- we change the logic that decides when a pipelinerun is finished. The following conditions should be met: the method that gets the next schedulable task does not return anything, no taskrun is in a transient state. We need to make sure that we don't meet this condition when for instance taskruns are waiting for a pod to be scheduled, but I think we should be safe there

Reply:
> I would like to propose a change in how we handle this logic here.
> I'm not sure I've thought enough about this, but these are the basic ideas:
> 
> * any taskrun is "permitted" to fail, it simply fails

not sure if I am following this, on a high level here not any taskrun is permitted to fail, after task is executed and if it fails, pipelinerun determines to exit with failure or continue executing rest of the tasks based on this flag `IsFailurePermitted`. This flag is set while parsing the pipelinerun.

> * we include failed taskruns in the list of executed tasks

Yes, we do include failed taskruns in the list of executed tasks.

> * we include the filters defined by the `runOn` in the logic that calculates which task should be included in the list of schedulable ones at a certain point in time

Yup, based on the states specified in `runOn` and actual states of a task, a list of schedulable tasks is generated.

> * we change the logic that decides when a pipelinerun is finished. The following conditions should be met: the method that gets the next schedulable task does not return anything, no taskrun is in a transient state. We need to make sure that we don't meet this condition when for instance taskruns are waiting for a pod to be scheduled, but I think we should be safe there

Again, not sure, if I am following this, but the logic which decides if pipeline is finished has changed and now includes, (1) check on failed task if failure is permitted (2) skipped tasks due to conflicts are counted as skipped